ariadne/utils.py

```diff
@@ -0,0 +1,260 @@
+import torch
+import numpy as np
+import itertools
+import pandas as pd
+from copy import deepcopy
+from ariadne.tracknet_v2.metrics import point_in_ellipse
+import faiss
+import matplotlib.pyplot as plt
+from scipy.interpolate import make_interp_spline, BSpline
+import glob
+import os
+
+def cartesian_product_two_stations(df):
+    first_station = df[df['station'] == 0][['r', 'phi', 'z', 'track']]
+    first_station.columns = ['r_left', 'phi_left', 'z_left', 'track_left']
+    second_station = df[df['station'] == 1][['r', 'phi', 'z', 'track']]
+    second_station.columns = ['r_right', 'phi_right', 'z_right', 'track_right']
+    rows = itertools.product(first_station.iterrows(), second_station.iterrows())
+    df = pd.DataFrame(left.append(right) for (_, left), (_, right) in rows)
+    df_fakes = df[(df['track_left'] == -1) & (df['track_right'] == -1)]
+    df = df[(df['track_left'] != df['track_right'])]
+    df = pd.concat([df, df_fakes], axis=0)
+    df = df.sample(frac=1).reset_index(drop=True)
+    return df.reset_index(drop=True)
+
+def weights_update(model, checkpoint):
+    model_dict = model.state_dict()
+    pretrained_dict = checkpoint['state_dict']
+    real_dict = {}
+    for (k, v) in model_dict.items():
+        needed_key = None
+        for pretr_key in pretrained_dict:
+            if k in pretr_key:
+                needed_key = pretr_key
+                break
+        assert needed_key is not None, "key %s not in pretrained_dict %r!" % (k, pretrained_dict.keys())
+        real_dict[k] = pretrained_dict[needed_key]
+
+    model.load_state_dict(real_dict)
+    model.eval()
+    return model
+
+def load_data(input_dir, file_mask, n_samples):
```

Comment:
please, add a few comment lines, because it is unclear which data is loading

Reply:
Commented